File: cmd/xcodecli/timeout_policy.go
package main

import (
	"fmt"
	"time"
)

const (
	defaultToolsListRequestTimeout   = 60 * time.Second
	defaultToolInspectRequestTimeout = 60 * time.Second
	defaultAgentGuideRequestTimeout  = 60 * time.Second
	defaultAgentDemoRequestTimeout   = 60 * time.Second

	defaultToolCallReadRequestTimeout  = 60 * time.Second
	defaultToolCallWriteRequestTimeout = 120 * time.Second
	defaultToolCallLongRequestTimeout  = 30 * time.Minute
	defaultToolCallFallbackTimeout     = 5 * time.Minute
)

func defaultToolCallTimeout(toolName string) time.Duration {
	switch toolName {
	case "XcodeListWindows", "XcodeLS", "XcodeGlob", "XcodeRead", "XcodeGrep", "GetBuildLog", "GetTestList", "XcodeListNavigatorIssues":
		return defaultToolCallReadRequestTimeout
	case "XcodeUpdate", "XcodeWrite", "XcodeRefreshCodeIssuesInFile":
		return defaultToolCallWriteRequestTimeout
	case "BuildProject", "RunAllTests", "RunSomeTests":
		return defaultToolCallLongRequestTimeout
	default:
		return defaultToolCallFallbackTimeout
	}
}

func formatTimeoutDuration(value time.Duration) string {
	if value <= 0 {
		return "0s"
	}
	if value%time.Hour == 0 {
		return fmt.Sprintf("%dh", int(value/time.Hour))
	}
	if value >= 5*time.Minute && value%time.Minute == 0 {
		return fmt.Sprintf("%dm", int(value/time.Minute))
	}
	if value%time.Second == 0 {
		return fmt.Sprintf("%ds", int(value/time.Second))
	}
	return value.String()
}

File: Sources/XcodeCLICore/Shared/TimeoutPolicy.swift
import Foundation

public enum TimeoutPolicy {
    // Category defaults (in seconds)
    public static let readTimeout: TimeInterval = 60
    public static let writeTimeout: TimeInterval = 120
    public static let longRunningTimeout: TimeInterval = 30 * 60 // 30 minutes
    public static let fallbackTimeout: TimeInterval = 5 * 60 // 5 minutes

    /// Returns the default timeout for a given tool name.
    public static func defaultToolCallTimeout(toolName: String) -> TimeInterval {
        switch toolName {
        case "XcodeListWindows", "XcodeLS", "XcodeGlob", "XcodeRead", "XcodeGrep",
             "GetBuildLog", "GetTestList", "XcodeListNavigatorIssues", "DocumentationSearch":
            return readTimeout
        case "XcodeUpdate", "XcodeWrite", "XcodeRefreshCodeIssuesInFile":
            return writeTimeout
        case "BuildProject", "RunAllTests", "RunSomeTests":
            return longRunningTimeout
        default:
            return fallbackTimeout
        }
    }

    /// Format a timeout duration for display.
    public static func formatDuration(_ seconds: TimeInterval) -> String {
        if seconds <= 0 { return "0s" }
        let totalSeconds = Int(seconds)
        if totalSeconds >= 3600 && totalSeconds % 3600 == 0 {
            return "\(totalSeconds / 3600)h"
        }
        if totalSeconds >= 300 && totalSeconds % 60 == 0 {
            return "\(totalSeconds / 60)m"
        }
        if seconds == Double(totalSeconds) {
            return "\(totalSeconds)s"
        }
        return String(format: "%.1fs", seconds)
    }
}
